File: python/chaparral/fine_tuning/models/knowledge.py
from dataclasses import dataclass
from typing import List, Literal

@dataclass
class Quantity:
    amount: float
    units: str

@dataclass
class Value:
    value: str | int | bool | Quantity

@dataclass
class Facet:
    name: str
    value: Value

@dataclass
class VerbTense:
    tense: Literal["past", "present", "future"]

@dataclass 
class ActionParam:
    name: str
    value: Value

@dataclass
class Action:
    verbs: List[str]
    verbTense: VerbTense
    subjectEntityName: str = "none"
    objectEntityName: str = "none"
    indirectObjectEntityName: str = "none"
    params: List[str | ActionParam] | None = None
    subjectEntityFacet: Facet | None = None

@dataclass
class ConcreteEntity:
    name: str
    type: str
    facets: List[Facet] | None = None

@dataclass
class KnowledgeResponse:
    entities: List[ConcreteEntity]
    actions: List[Action]
    topics: List[str]
    inverseActions: List[Action] | None = None

File: dotnet/typeagent/src/knowpro/Knowledge.cs
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

using System.Text.Json.Serialization;

namespace TypeAgent.KnowPro;

[JsonPolymorphic(TypeDiscriminatorPropertyName = "$type")]
[JsonDerivedType(typeof(ConcreteEntity), "entity")]
[JsonDerivedType(typeof(Action), "action")]
[JsonDerivedType(typeof(Topic), "topic")]
[JsonDerivedType(typeof(Tag), "tag")]
[JsonDerivedType(typeof(StructuredTag), "structuredTag")]
public abstract class Knowledge
{
    public Knowledge() { }

    [JsonIgnore]
    public abstract KnowledgeType KnowledgeType { get; }
}

public interface IKnowledgeSource
{
    KnowledgeResponse? GetKnowledge();
}


public partial class ConcreteEntity : Knowledge
{
    [JsonPropertyName("name")]
    public string Name { get; set; }

    [JsonPropertyName("type")]
    public string[] Type { get; set; }

    [JsonPropertyName("facets")]
    public Facet[]? Facets { get; set; }
}

public partial class Action : Knowledge
{
    [JsonPropertyName("verbs")]
    public string[] Verbs { get; set; }

    [JsonPropertyName("verbTense")]
    public string VerbTense { get; set; }

    [JsonPropertyName("subjectEntityName")]
    public string SubjectEntityName { get; set; }

    [JsonPropertyName("objectEntityName")]
    public string ObjectEntityName { get; set; }

    [JsonPropertyName("indirecObjectEntityName")]
    public string IndirectObjectEntityName { get; set; }

    [JsonPropertyName("params")]
    public IActionParam[]? Params { get; set; }
}

public class KnowledgeResponse
{
    [JsonPropertyName("entities")]
    [JsonRequired]
    public ConcreteEntity[] Entities { get; set; }

    [JsonPropertyName("actions")]
    [JsonRequired]
    public Action[] Actions { get; set; }

    [JsonPropertyName("inverseActions")]
    public Action[]? InverseActions { get; set; }

    [JsonPropertyName("topics")]
    public string[] Topics { get; set; }
}

public partial class Topic : Knowledge
{
    [JsonPropertyName("text")]
    public string Text { get; set; }
}

public partial class Tag : Knowledge
{
    [JsonPropertyName("text")]
    public string Text { get; set; }
}

public partial class StructuredTag : ConcreteEntity
{

}
